### DISCUSSION TITLE
Disconnect for Lambda?

### DISCUSSION THREAD

**Discussion by jonthewayne** [prisma/prisma#2221] [Q&A] [answered]
The Prisma docs mention that it is recommended to always explicitly call prisma.disconnect() in our code. Your quickstart example shows that, but when I checked out your aws lambda example, it didn't show it. 

https://github.com/prisma/prisma-examples/tree/master/deployment-platforms/aws-lambda

We're deploying to AWS lambda and I wanted to double check with you what we should do.

We're actually using graphql-yoga server in our lambda function, so it'd be wonderful to have an example for how to add this to the server invocation in some way. 

Thank you!

**Comment by 2color** [comment]
Hey @jonthewayne,

That recommendation applies in the context of a long-running Node.js process that uses Prisma Client.

**In the context of AWS Lambda, and more broadly serverless environment, it's ok not disconnect so that further invocations of the Serverless function can reuse the DB connection.** 

It's worth noting that reusing the connection, i.e. omitting the disconnect call, in Serverless functions has its limits, as it can exhaust the connection pool depending on your traffic pattern and concurrency limit (the number of serverless functions that can be invoked simultaneously). 

For a more elaborate explanation of the considerations, check out the [Serverless deployment docs](https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-client/deployment#serverless-faas)

**Reply by jonthewayne** [reply]
Thanks Daniel!

**Comment by jonthewayne** [comment]
I can confirm that when using Graphql Yoga server, and passing in `prisma` via the server context,  it will hang and spit back a 502 internal server error unless I actually use `prisma.disconnect()` in the resolver. If I use it, function is super quick, works repeatedly.

I tried creating a middleware that would call `context.prisma.disconnect()` after every resolver, but it must hang before this point, because that didn't work.

**Reply by 2color** [reply]
Link to the related issue: https://github.com/prisma/prisma/issues/2241

**Comment by 2color** [comment, marked as answer]
The fix for this, as detailed in https://github.com/prisma/prisma/issues/2241#issuecomment-616923675 is to set the the `context.callbackWaitsForEmptyEventLoop` of the [AWS Lambda Context Object](https://docs.aws.amazon.com/lambda/latest/dg/nodejs-context.html) to `false` as follows:

```js
exports.server = (event, context, cb) => {
  // Set to false to send the response right away when the callback executes, instead of waiting for the Node.js event loop to be empty.
  context.callbackWaitsForEmptyEventLoop = false

  return lambda.graphqlHandler(event, context, cb)
}
```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
